src/Compilers/CSharp/Portable/Symbols/NativeIntegerTypeSymbol.cs

```diff
@@ -93,50 +172,239 @@ internal NativeIntegerTypeSymbol(NamedTypeSymbol underlyingType) : base(underlyi
+        {
+            Debug.Assert(!nativeIntegerSymbol.Equals(underlyingSymbol, TypeCompareKind.ConsiderEverything));
+            Debug.Assert(!underlyingSymbol.Equals(nativeIntegerSymbol, TypeCompareKind.ConsiderEverything));
+            Debug.Assert(nativeIntegerSymbol.Equals(underlyingSymbol, TypeCompareKind.IgnoreNativeIntegers));
+            Debug.Assert(underlyingSymbol.Equals(nativeIntegerSymbol, TypeCompareKind.IgnoreNativeIntegers));
+            Debug.Assert(nativeIntegerSymbol.GetHashCode() == underlyingSymbol.GetHashCode());
+        }
+
+        private sealed class NativeIntegerTypeMap : AbstractTypeMap
+        {
+            private readonly NativeIntegerTypeSymbol _type;
+            private readonly SpecialType _specialType;
+
+            internal NativeIntegerTypeMap(NativeIntegerTypeSymbol type)
+            {
+                _type = type;
+                _specialType = _type.UnderlyingNamedType.SpecialType;
+
+                Debug.Assert(_specialType == SpecialType.System_IntPtr || _specialType == SpecialType.System_UIntPtr);
+            }
+
+            internal override NamedTypeSymbol SubstituteTypeDeclaration(NamedTypeSymbol previous)
+            {
+                return previous.SpecialType == _specialType ?
+                    _type :
+                    base.SubstituteTypeDeclaration(previous);
+            }
+        }
+    }
+
+    internal sealed class NativeIntegerMethodSymbol : WrappedMethodSymbol, Cci.IReference
+    {
+        private readonly NativeIntegerTypeSymbol _container;
+        private readonly Symbol? _associatedSymbol;
+
+        internal NativeIntegerMethodSymbol(NativeIntegerTypeSymbol container, MethodSymbol underlyingMethod, Symbol? associatedSymbol)
+        {
+            Debug.Assert(!underlyingMethod.IsGenericMethod);
+            _container = container;
+            _associatedSymbol = associatedSymbol;
+            UnderlyingMethod = underlyingMethod;
+            NativeIntegerTypeSymbol.VerifyEquality(this, underlyingMethod);
+        }
+
+        public override Symbol ContainingSymbol => _container;
+
+        public override MethodSymbol UnderlyingMethod { get; }
+
+        public override TypeWithAnnotations ReturnTypeWithAnnotations => _container.SubstituteUnderlyingType(UnderlyingMethod.ReturnTypeWithAnnotations);
+
+        public override ImmutableArray<TypeWithAnnotations> TypeArgumentsWithAnnotations => ImmutableArray<TypeWithAnnotations>.Empty;
 
-                return ImmutableArray<NamedTypeSymbol>.Empty;
+        public override ImmutableArray<TypeParameterSymbol> TypeParameters => ImmutableArray<TypeParameterSymbol>.Empty;
+
+        public override ImmutableArray<ParameterSymbol> Parameters =>
+            UnderlyingMethod.Parameters.SelectAsArray((Func<ParameterSymbol, NativeIntegerMethodSymbol, ParameterSymbol>)((p, m) => (ParameterSymbol)new NativeIntegerParameterSymbol(m._container, m, p)), this);
```

Comment:
>UnderlyingMethod.Parameters.SelectAsArray((Func<ParameterSymbol, NativeIntegerMethodSymbol, ParameterSymbol>)((p, m) => (ParameterSymbol)new NativeIntegerParameterSymbol(m._container, m, p)), this) [](start = 12, length = 197)

It feels like we should cache these. Consumers, usually, do not expect getting different instance for parameters every time.